File: PythonImplementation/Sort/MergeSort.py
import SortTestHelper


def __merge(arr , l, mid, r):

    #归并操作，arr为原数组, lr 为带排序的部分索引值
    #k为在原数组上的待排序位置
    #ij用于标记在辅助数组中左右两部分相比较的索引值

    assist = arr[l:r+1]
    k = l
    i = l
    j = mid+1

    for k in range(l, r+1):
        #先检查索引范围，当排序未结束时，及k未在lr之间遍历完时
        #如果i j超过了各自的部分，则将另一部分的剩余部分一次填入原数组
        if(i > mid):
            arr[k] = assist[j-l]
            j+=1
        elif(j > r):
            arr[k] = assist[i-l]
            i+=1
        else:
            if(assist[i-l] <= assist[j-l]):
                arr[k] = assist[i-l]
                i+=1
            else:
                arr[k] = assist[j-l]
                j+=1





def __mergesort(arr, l, r):

    #递归调用，将两部分排序好之后再归并成一组


    if(l >= r):
        return

    mid = int((l+r)/2)
    __mergesort(arr, l, mid)
    __mergesort(arr, mid+1, r)
    __merge(arr, l, mid, r)

def mergesort(arr):
    __mergesort(arr, 0, int(len(arr))-1 )
    return arr

File: JavaIplementation/src/SortCollection/MergeSort.java
package SortCollection;

import Helper.SortTestHelper;

import java.util.Arrays;

public class MergeSort {
    //The entrance of the merge sort function
    public void Merge(int[] list){
        _Mergesort(list, 0, list.length -1);
    }
    //merge function : merge sort the specific interval.
    private void _Mergesort(int[] list, int l, int r){
        //list : list to be ordered
        //l : the left border of the to-be-ordered interval.
        //r : the right border of the to-be-ordered interval.
        if (l >= r)
            return;


        int mid = (l + r) / 2;
        _Mergesort(list, l, mid);
        _Mergesort(list, mid+1, r);
        _Merge(list, l, mid, r);

    }

    //merge two part in a list into an ordered one .
    private void _Merge(int[] list, int l, int mid, int r){
       
        int [] assist =  Arrays.copyOfRange(list, l, r+1);
        int k = l;
        int i = l;
        int j = mid + 1;

        for (;k <= r; k++){
            if(i > mid){
                list[k] = assist[j-l];
                j++;
            }
            else if(j > r){
                list[k] = assist[i-l];
                i++;
            }
            else {
                if (assist[i-l] <= assist[j-l]){
                    list[k] = assist[i-l];
                    i++;
                }
                else{
                    list[k] = assist[j-l];
                    j++;
                }
            }

        }

    }

    public static void main(String[] args){
        int [] a = {6,1,2,2,4,6,8,33,1,43};
        MergeSort mergeSort = new MergeSort();
        mergeSort.Merge(a);
        System.out.println(Arrays.toString(a));
    }

    public void SortTest(int n){
        //n for the size of the list
        //the item varies from 0 to n.
        SortTestHelper helper = new SortTestHelper();
        int[] list = helper.RadnomTestList(n, 0, n);

        System.out.println("The original list is:");
        for (int i =0;i < list.length;i++){
            System.out.printf(" %d",list[i]);
        }
        System.out.printf("\n");
        double startTime = System.currentTimeMillis();
        Merge(list);
        double endTime = System.currentTimeMillis();


        System.out.println("The ordered list is:");
        for (int i = 0;i < list.length; i++){
            System.out.printf(" %d",list[i]);
        }

        System.out.println("\nSortType: Merge Sort");
        System.out.println("Time:" + (endTime - startTime)/1000 + " second");

    }

    public double getTime(int n){
        //get the time for sorting an array of length n
        SortTestHelper helper = new SortTestHelper();
        int[] list = helper.RadnomTestList(n, 0, n);

        double startTime = System.currentTimeMillis();
        Merge(list);
        double endTime = System.currentTimeMillis();
        double second = (endTime - startTime)/1000;

        System.out.println("Sort type: Merge Sort");
        System.out.println("Size:" + n);
        System.out.println("Time: "+ second + "second");
        return second;
    }

}
